nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -707,12 +736,41 @@ impl DataStore {
                             // delete a read-only downstairs running for a
                             // snapshot that doesn't exist will return a 404,
                             // causing the saga to error and unwind.
+                            //
+                            // XXX are any other filters required, except the
+                            // three from snapshots_to_delete?
                             .filter(
                                 dsl::snapshot_addr.eq_any(
                                     &crucible_targets.read_only_targets,
                                 ),
                             )
-                            .filter(dsl::volume_references.eq(0))
+                            // XXX is there a better way to do this, rather than
+                            // one filter per struct field?
```

Comment:
Is it possible this could be incorrect (in addition to being awkward, which I assume is the reason for the comment)? E.g., if `snapshots_to_delete` contained the dataset_id/region_id/snapshot_id triples

```
(A, B, C),
(D, E, F),
```

and the database contained a row with `(A, B, F)`, then the three independent filters would all be true even though `(A, B, F)` itself is not one of the triples we're trying to delete. I _think_ such a contrived example is not actually possible, but I'm not entirely sure?

I'm not sure how to rework this, particularly in diesel. In SQL terms I think we'd want something like a subquery containing the `snapshots_to_delete` (instead of having them in a `Vec`) so that we could do a `WHERE EXISTS ...` on the three columns we care about, maybe? If we want to keep `snapshots_to_delete` in Rust, how bad would it be to loop over them and run this query once for each of them?

Reply:
Turns out I can chain them together with `.and`: 042b6125

Reply:
I think that still has the problem of `snapshots_to_delete = (A,B,C), (D,E,F)` matching `(A,B,F)`, right? But maybe that's fine / impossible in practice? It is an improvement over the chained filters for sure though.